Fix the issue that all data on the report pages are not shown

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #1183

- Make `path` argument of `chain` and `set` functions accept array type to fix the incorrect path parsing issue of report data in the reducer.
- Extra change: Change the dynamic storing paths of MC product feed to use array `path` to make it look clearer in the reducer.
- Extra change: Add the **.externalized.json** file to the ignored watching patterns of jest to avoid repeatedly running the same test after modifying js file.

### Screenshots:

![image](https://user-images.githubusercontent.com/17420811/148714572-aacfc137-2c58-4c62-80b1-9a686e63c539.png)

### Detailed test instructions:

#### The report page

1. Add snippets in the **google-listings-and-ads.php** file if you don't have report data to check it.
    ```php
    add_filter(
    	'woocommerce_gla_prepared_response_mc-reports-programs',
    	function( $response ) {
    		return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/programs.json' ), true ) ?: [];
    	},
    );
    
    add_filter(
    	'woocommerce_gla_prepared_response_ads-reports-programs',
    	function( $response ) {
    		return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
    	},
    );
    ``` 
2. Go to the programs report page `/wp-admin/admin.php?page=wc-admin&reportKey=programs&path=%2Fgoogle%2Freports`.
3. Data in the chart and **Program** table should be displayed.

#### The product feed page

1. Go to the programs report page `/wp-admin/admin.php?page=wc-admin&reportKey=programs&path=%2Fgoogle%2Fproduct-feed`.
2. Data in the **Issues to resolve** table should be displayed as before.

### Changelog entry

eason9487 committed:
Make `path` argument of `chain` and `set` functions accept array type to fix the incorrect path parsing issue of report data in reducer.

eason9487 committed:
Change the dynamic storing path of MC product feed to use array paths to make it look clearer.

eason9487 committed:
Don't watch the .externalized.json file change for the jest tests

ianlin commented on js/src/data/reducer.js:
```diff
@@ -46,10 +46,19 @@ function chain( state, basePath = '' ) {
 		}
 		return clone( value );
 	};
+	const combinBasePath = ( path ) => {
+		if ( basePath ) {
+			if ( Array.isArray( basePath ) || Array.isArray( path ) ) {
+				return [].concat( basePath, path );
+			}
+			return `${ basePath }.${ path }`;
+		}
+		return path;
+	};
```
💅  This function returns a `string` or an `array` and it took me a while to understand it's because lodash `setWith` 's second parameter `path` can be either a string or an array. Might be better to add a comment here 😃

eason9487 replied:
Added by https://github.com/woocommerce/google-listings-and-ads/pull/1184/commits/eb552c08e0536aca8a76718651d47997cafb8ec3.

ianlin reviewed (approved):
Tested locally and everything works as expected. 👍 

✅  I can reproduce it using the `develop` branch, none is rendered in the reports page.
✅  Using this PR's branch, I can see the data rendered in the chart and the Program table.
✅  Using this PR's branch, I can see the **Issues to resolve table** is displayed normally.
✅  The newly added unit test can catch the error using the code in `develop` branch.

jorgemd24 commented on js/src/data/reducer.js:
```diff
@@ -46,10 +46,19 @@ function chain( state, basePath = '' ) {
 		}
 		return clone( value );
 	};
+	const combinBasePath = ( path ) => {
```
❔

Is it needed to put this function inside of the `chain function` scope? I think we could just have this function outside of the `chain function` scope receiving two params: `basePath` and `path` thus it could be re-used easily in the future.

💅  I think there is a misspelling with the function's name `combinBasePath` should be `combineBasePath`

eason9487 replied:
Thanks for catching the typo. Fixed by https://github.com/woocommerce/google-listings-and-ads/pull/1184/commits/6534674b77c9cf934f7467e93b15d17c58485c41.

> Is it needed to put this function inside of the `chain function` scope? I think we could just have this function outside of the `chain function` scope receiving two params: `basePath` and `path` thus it could be re-used easily in the future.

I guess its reuse is likely to be relatively low. I'd prefer to move it out when the needs happen.

jorgemd24 replied:
Thanks for correcting this! 

> I guess its reuse is likely to be relatively low. I'd prefer to move it out when the needs happen.

Let's keep it as it is so we do not delay this PR and if in the future is needed we can just move it out.

jorgemd24 commented on js/src/data/reducer.js:
```diff
@@ -46,10 +46,19 @@ function chain( state, basePath = '' ) {
+	};
 
 	return {
 		set( path, value ) {
```
❔ 

This is not related to this PR but sometimes I get confused with the `set function` that has the same name as [lodash set](https://lodash.com/docs/4.17.15#set) and it can be easily mistaken.

eason9487 replied:
Thanks for pointing out this. What about changing `set` to `setIn`, and `chain` to `chainState`? And I lean toward renaming them in a separate RP.

jorgemd24 replied:
Yes, those names are better and it will help to avoid any confusion with lodash. Thanks!

eason9487 replied:
The follow-up PR: #1191

eason9487 committed:
Fix the misspelling function name of `combinBasePath`.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1184#discussion_r781178361

eason9487 committed:
Add JSDoc for `chain` and `set` in the reducer.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1184#discussion_r780979196
